### PR TITLE
CNF-7610: Mixed CPUs plugin deployment support

### PR DESCRIPTION
**Pull request by Tal-or**
This PR contains set of changes that are needed for MixedCPUs feature support:

- [x]  Bootstrap the new API used for the feature activation 
- [x] Append the shared cpus as part of Kubelet's reservedSystemCPUs
- [x] Deploy all the different configuration files that are needed for the mixed-cpus feature activation
- [x] Unit tests

The feature is optional and off by default.

E2E tests will be added in a separate PR since this PR alone doesn't provide any added functionality and depends on other changes.

Related to: 
* https://github.com/openshift/kubernetes/pull/1795
* https://github.com/openshift/kubernetes/pull/1799
* https://github.com/cri-o/cri-o/pull/7502
* #858 

EP: https://github.com/openshift/enhancements/commit/76ea48a88bca15ef6157aacf90b8544a763e6075

### CODE REVIEW

File: pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go [hunk 1]
```diff
@@ -9,9 +9,11 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go:12]: let's use the `k8s.io/utils` import

**Review by ffromani — COMMENTED** [review]
Did we add validation already? we need to check how the shared pool composes with other pools. E.g. (purely random thoughts): not overlap with isolated or offlined, potentially a subset of reserved? not overlapping ALSO with reserved?

File: pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go (hunk 1 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go:12]: Done. Thanks!

**Comment by Tal-or** [comment]
> Did we add validation already? we need to check how the shared pool composes with other pools. E.g. (purely random thoughts): not overlap with isolated or offlined, potentially a subset of reserved? not overlapping ALSO with reserved?

It should not overlap with any of the sets.
The fact that we're merging reserved + shared later on under Kubelet's systemReserved is an implementation detail.

**Comment by Tal-or** [comment]
depends on #858

**Comment by Tal-or** [comment]
/test e2e-gcp-pao 
Flaky

File: test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			workers, err := nodes.GetByLabels(testutils.NodeSelectorLabels)
+			Expect(err).ToNot(HaveOccurred())
+			// test arbitrary one should be good enough
+			worker := &workers[0]
+			cmd := []string{
+				"chroot",
+				"/rootfs",
+				"/bin/bash",
+				"-c",
+				fmt.Sprintf("/bin/grep -E 'shared_cpuset.*' %s | awk -F '\"' '{ print $2 }'", crioRuntimeConfigFile),
```
Comment (ffromani) [inline test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go:98]: minor (for a future improvement):
1. if you use the absolute path for grep (good), why not for awk?
2. you can just call awk, no need for a pipe

File: test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	return teardown
+}
+
+func cpuSetToPerformanceCPUSet(set *cpuset.CPUSet) *performancev2.CPUSet {
+	c := performancev2.CPUSet(set.String())
+	return &c
+}
+
+func isFileExistCmd(absoluteFileName string) []string {
```
Comment (ffromani) [inline test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go:152]: looks correct. A stronger test could be try to read the file and ensure it has _any_ content on it. Is not much costlier than the current test. But it's debatable we need it on this PR, so let's save for the future.

File: pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go [hunk 4]
```diff
@@ -98,6 +99,18 @@ func New(profile *performancev2.PerformanceProfile, profileMCPLabels map[string]
...
+		if err != nil {
+			return nil, err
+		}
+		reservedCPUs, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
+		if err != nil {
+			return nil, err
+		}
+		kubeletConfig.ReservedSystemCPUs = reservedCPUs.Union(sharedCPUs).String()
+	}
+
```
Comment (swatisehgal) [inline pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go:113]: Just wondering if we should be doing some sort validation to ensure that the CPUs are non-overlapping for `shared` and `reserved` fields?

File: pkg/apis/performanceprofile/v2/performanceprofile_types.go [hunk 5]
```diff
@@ -114,6 +114,11 @@ type CPU struct {
 	// Offline defines a set of CPUs that will be unused and set offline
 	// +optional
 	Offlined *CPUSet `json:"offlined,omitempty"`
+	// Shared defines a set of CPUs that will be shared among guaranteed workloads
+	// that needs additional cpus which are not exclusive,
+	// alongside the isolated, exclusive resources that are being used already by those workloads.
+	// +optional
+	Shared *CPUSet `json:"shared,omitempty"`
```
Comment (swatisehgal) [inline pkg/apis/performanceprofile/v2/performanceprofile_types.go:121]: Hmm, we are introducing the shared field here but the previous commit already uses it. I think we should reorder the commits maybe?

File: pkg/performanceprofile/controller/performanceprofile/components/components.go [hunk 6]
```diff
@@ -27,10 +27,12 @@ type Options struct {
 }
```
Comment (swatisehgal) [inline pkg/performanceprofile/controller/performanceprofile/components/components.go:27]: The commit message does not match the implementation, for example appending of the `shared` CPUs to the `reservedSystemCPUs` is not happening here and happened earlier. Am I missing something here?

**Review by swatisehgal — COMMENTED** [review]
Looks promising, just needs some minor polishing to ensure that the commits are ordered correctly and there is no commit message and implementation mismatch.

File: test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go (hunk 3 above)
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go]: I replace `-f` with `-s` to check if file exist and not empty. Thanks!

File: test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go (hunk 2 above)
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/10_mixedcpus/mixedcpus.go]: TIL

File: test/e2e/performanceprofile/functests/10_mixedcpus/10_mixedcpus_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package __mixedcpus_test
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/10_mixedcpus/10_mixedcpus_suite_test.go]: Changed to `11` thanks

File: pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go (hunk 4 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/components/kubeletconfig/kubeletconfig.go:113]: Hey @swatisehgal Thank you for taking the time to review the PR.
We do perform validation during the profile application using admission hooks.
see:
https://github.com/openshift/cluster-node-tuning-operator/pull/853/commits/a36f6f027eb2750b14ab2a111cd4ce3c8b2e6c20#diff-1a001c69ad51c074896e9bc0c708d605c59f65bd4fe82dd95a6f882358ac82f8R160

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	Expect(err).ToNot(HaveOccurred())
+
+	// arbitrary take the first one
+	sharedcpu := cpuset.New(isolated.List()[0])
+	testlog.Infof("shared cpu ids are: %q", sharedcpu.String())
+	updatedIsolated := isolated.Difference(sharedcpu)
+	profile.Spec.CPU.Isolated = cpuSetToPerformanceCPUSet(&updatedIsolated)
+	profile.Spec.CPU.Shared = cpuSetToPerformanceCPUSet(&sharedcpu)
+	profile.Spec.WorkloadHints.MixedCpus = pointer.Bool(true)
+	Expect(testclient.Client.Update(ctx, profile)).ToNot(HaveOccurred(), "failed to update profile: %q", profile.Name)
```
Comment (shajmakh) [inline test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go:125]: do we want to update the profile in a loop?
https://github.com/openshift/cluster-node-tuning-operator/blob/master/test/e2e/performanceprofile/functests/utils/profiles/profiles.go#L101

File: pkg/apis/performanceprofile/v2/performanceprofile_types.go (hunk 5 above)
Comment (Tal-or) [reply pkg/apis/performanceprofile/v2/performanceprofile_types.go:121]: There was a huge mess with the first commit you're absolutely right. I fixed it. Thanks!

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+	mcp, err := mcps.GetByProfile(profile)
+	Expect(err).ToNot(HaveOccurred())
+
+	mcps.WaitForCondition(mcp, machineconfigv1.MachineConfigPoolUpdating, corev1.ConditionTrue)
+	mcps.WaitForCondition(mcp, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
+
```
Comment (shajmakh) [inline test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go:132]: here and in the returned teardown function, you can consider using this util function:
https://github.com/openshift/cluster-node-tuning-operator/blob/master/test/e2e/performanceprofile/functests/utils/profilesupdate/profile_update.go#L43,  it will make sure that the profile got updated as expected.

**Review by shajmakh — COMMENTED** [review]
Thanks for the PR, tests looks good to me.
just a minor comment.

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go (hunk 8 above)
Comment (ffromani) [reply test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go:125]: yes please, thanks for catching this. Updates can fail (commonly) and we should always retry them/wrap them in a loop.

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go (hunk 9 above)
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go:132]: @shajmakh What is the different then between the UpdateWithRetry and this one?

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+			// don't need to check the error, the values were already validated.
+			shared, _ := cpuset.Parse(string(*profile.Spec.CPU.Shared))
+			Expect(shared.Equals(crioShared)).To(BeTrue(), "crio config file does not contain the expected shared cpuset; shared=%q crioShared=%q",
+				shared.String(), crioShared.String())
+		})
+	})
+})
+
+func setup(ctx context.Context, profile *performancev2.PerformanceProfile) func(ctx context.Context) {
+	initialProfile := profile.DeepCopy()
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go:114]: I need to investigate exactly how to do that, but anyway we should enable new techpreview job on which those tests will be running and on which FG would be enable by default.

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go (hunk 8 above)
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go]: Done. Thanks!

File: test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	Expect(testclient.Client.Update(ctx, profile)).ToNot(HaveOccurred(), "failed to update profile: %q", profile.Name)
+
+	mcp, err := mcps.GetByProfile(profile)
+	Expect(err).ToNot(HaveOccurred())
+
+	mcps.WaitForCondition(mcp, machineconfigv1.MachineConfigPoolUpdating, corev1.ConditionTrue)
+	mcps.WaitForCondition(mcp, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
+
+	teardown := func(ctx context.Context) {
+		Expect(testclient.Client.Get(ctx, client.ObjectKeyFromObject(initialProfile), profile))
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/11_mixedcpus/mixedcpus.go:136]: Done. Thanks

**Comment by Tal-or** [comment]
Need to reprovision my cluster to validate FG integration

**Comment by Tal-or** [comment]
```
  Unexpected error:
      <wait.errInterrupted>: 
      timed out waiting for the condition
      {
          cause: <*errors.errorString | 0xc0002bd6b0>{
              s: "timed out waiting for the condition",
          },
      }
  occurred
  In [It] at: /go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/1_performance/cpu_management.go:298 @ 12/07/23 11:28:00.169
```
The lane is falling consistently maybe the issue is real. working to figure it out

**Comment by ffromani** [comment]
> ```
>   Unexpected error:
>       <wait.errInterrupted>: 
>       timed out waiting for the condition
>       {
>           cause: <*errors.errorString | 0xc0002bd6b0>{
>               s: "timed out waiting for the condition",
>           },
>       }
>   occurred
>   In [It] at: /go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/1_performance/cpu_management.go:298 @ 12/07/23 11:28:00.169
> ```
> 
> The lane is falling consistently maybe the issue is real. working to figure it out

```
,Reason:Unschedulable,Message:0/6 nodes are available: 1 Insufficient cpu, 2 node(s) didn't match Pod's node affinity/selector, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }. preemption: 0/6 nodes are available: 1 No preemption victims found for incoming pod, 5 Preemption is not helpful for scheduling..,},},Message:,Reason:,HostIP:,PodIP:,StartTime:<nil>,ContainerStatuses:[]ContainerStatus{},QOSClass:Guaranteed,InitContainerStatuses:[]ContainerStatus{},NominatedNodeName:,PodIPs:[]PodIP{},EphemeralContainerStatuses:[]ContainerStatus{},Resize:,},}
```
suggests there could be a pod being deleted and not yet gone, or that in general the cluster state is polluted somehow when the tests runs.

**Comment by Tal-or** [comment]
There's an issue with the PR:
We added a new annotation to crio under the allowed annotation section:
https://github.com/openshift/cluster-node-tuning-operator/pull/853/commits/2375df36f2b91615367a528e932b4572a345a72c#diff-e763c8f403f44cd298b79a200633f6e687c478a00bda33b0d85e0add10eeb60bR21 

but since the existing CRIO version that is running in the CI is not familiar with this new annotation (PR is till open: https://github.com/cri-o/cri-o/pull/7502) it would failed to validate the `high-performance` runtime and hence will ignore it:  https://github.com/cri-o/cri-o/blob/1530af133d95de9cf4e5502c2f7dba4faa286a04/pkg/config/config.go#L1235 
Missing the `high-performance` runtime configuration will cause the test the request for lb disable to fail

Bottom line:
We should merge CRIO patch first.

**Comment by ffromani** [comment]
[...]
> Bottom line: We should merge CRIO patch first.

...and let the CI infra consume the updated package once released. Is that feasible under the current timeline constraints?

File: assets/performanceprofile/configs/99-runtimes.conf [hunk 12]
```diff
@@ -17,4 +18,4 @@ runtime_root = "/run/runc"
 runtime_path = "{{.RuntimePath}}"
 runtime_type = "oci"
 runtime_root = "{{.RuntimeRoot}}"
-allowed_annotations = ["cpu-load-balancing.crio.io", "cpu-quota.crio.io", "irq-load-balancing.crio.io", "cpu-c-states.crio.io", "cpu-freq-governor.crio.io"]
+allowed_annotations = ["cpu-load-balancing.crio.io", "cpu-quota.crio.io", "irq-load-balancing.crio.io", "cpu-c-states.crio.io", "cpu-freq-governor.crio.io", "cpu-shared.crio.io"]
```
Comment (mrunalp) [inline assets/performanceprofile/configs/99-runtimes.conf:21]: This should be included in the list only if Feature Gate is enabled.

**Comment by Tal-or** [comment]
> There's an issue with the PR: We added a new annotation to crio under the allowed annotation section: [2375df3#diff-e763c8f403f44cd298b79a200633f6e687c478a00bda33b0d85e0add10eeb60bR21](https://github.com/openshift/cluster-node-tuning-operator/commit/2375df36f2b91615367a528e932b4572a345a72c#diff-e763c8f403f44cd298b79a200633f6e687c478a00bda33b0d85e0add10eeb60bR21)
> 
> but since the existing CRIO version that is running in the CI is not familiar with this new annotation (PR is till open: [cri-o/cri-o#7502](https://github.com/cri-o/cri-o/pull/7502)) it would failed to validate the `high-performance` runtime and hence will ignore it: https://github.com/cri-o/cri-o/blob/1530af133d95de9cf4e5502c2f7dba4faa286a04/pkg/config/config.go#L1235 Missing the `high-performance` runtime configuration will cause the test the request for lb disable to fail
> 
> Bottom line: We should merge CRIO patch first.

I removed the annotation in case the FG disabled. This should remove the change from the `allowed_annotation` configured for crio and now tests should pass with no issues.
